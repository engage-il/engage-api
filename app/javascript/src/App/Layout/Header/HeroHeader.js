// @flow
import * as React from 'react'
import { findDOMNode as findDomNode } from 'react-dom'
import styled from 'react-emotion'
import { HeaderNavbar } from './HeaderNavbar'
import { Logo } from '../Logo'
import { mixins } from '@/styles'

type State = {
  isSticky: boolean
}

export class HeroHeader extends React.Component<*, *, State> {
  state = {
    isSticky: false
  }

  // refs
  navbar: ?Object = null

  // actions
  setNavbar = (navbar: ?Object) => {
    this.navbar = navbar
    navbar && this.invalidateIsSticky(navbar)
  }

  invalidateIsSticky = (navbar: ?Object) => {
    this.setState(() => {
      const node = findDomNode(navbar)
      const parent = node && node.parentElement
      const isSticky = parent && parent.getBoundingClientRect().top < 0
      return { isSticky }
    })
  }

  // events
  didScroll = () => {
    this.invalidateIsSticky(this.navbar)
  }

  // lifecycle
  componentDidMount () {
    window.addEventListener('scroll', this.didScroll)
  }

  componentWillUnmount () {
    window.removeEventListener('scroll', this.didScroll)
  }

  render () {
    const { isSticky } = this.state

    return (
      <Container>
        <Hero>
          <Logo width={58} height={49} />
          <h1>Legislated</h1>
        </Hero>
        <HeaderNavbar
          isSticky={isSticky}
          navbarRef={this.setNavbar}
        />
      </Container>
    )
  }
}

const Container = styled.header`
  ${mixins.flexColumn};

  align-items: center;
  padding-top: 70px;

  ${mixins.mobile`
    display: none;
  `}
`

const Hero = styled.div`
  ${mixins.flexColumn};

  align-items: center;
  margin-bottom: 30px;

  > img {
    margin-bottom: 15px;
  }
`
