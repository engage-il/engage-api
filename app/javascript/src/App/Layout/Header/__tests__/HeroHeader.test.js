/* eslint-env jest */
import * as React from 'react'
import { shallow } from 'enzyme'
import { findDOMNode as findDomNode } from 'react-dom'
import { HeroHeader } from '../HeroHeader'

// mocks
jest.mock('react-dom')

const getBoundingClientRect = jest.fn()

findDomNode.mockReturnValue({
  parentElement: {
    getBoundingClientRect
  }
})

// subject
let subject

function loadSubject () {
  subject = shallow(<HeroHeader />)
}

// spec
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows the sticky navbar after scrolling', () => {
    loadSubject()
    subject.setState({ isSticky: true })
    expect(subject).toMatchSnapshot()
  })
})

describe('scrolling', () => {
  it('stickies the navbar once it scrolls below the top of the page', () => {
    loadSubject()
    getBoundingClientRect.mockReturnValueOnce({ top: -1 })
    subject.instance().didScroll()
    expect(subject).toHaveState('isSticky', true)
  })

  it('unstickies the navbar once it scrolls above the top of the page', () => {
    loadSubject()
    subject.setState({ isSticky: true })
    getBoundingClientRect.mockReturnValueOnce({ top: 1 })
    subject.instance().didScroll()
    expect(subject).toHaveState('isSticky', false)
  })
})
