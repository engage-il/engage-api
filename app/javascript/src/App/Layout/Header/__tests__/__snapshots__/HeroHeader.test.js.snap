// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows the sticky navbar after scrolling 1`] = `
<Styled(header)>
  <Styled(div)>
    <Logo
      height={49}
      width={58}
    />
    <h1>
      Legislated
    </h1>
  </Styled(div)>
  <HeaderNavbar
    isSticky={true}
    navbarRef={[Function]}
  />
</Styled(header)>
`;
